src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj

```diff
@@ -90,13 +95,25 @@
    [this env-name]
    (let [environment-class-info (:environment-class-info-tags
                                  (tk-services/service-context this))]
-     (get @environment-class-info env-name)))
+     (get-in @environment-class-info [env-name :tag])))
+
+  (get-environment-class-info-tag-last-updated
+   [this env-name]
+   (let [environment-class-info (:environment-class-info-tags
+                                 (tk-services/service-context this))]
+     (get-in @environment-class-info [env-name :last-updated])))
 
   (set-environment-class-info-tag!
-   [this env-name tag]
+   [this env-name tag last-update-before-tag-computed]
    (let [environment-class-info (:environment-class-info-tags
                                  (tk-services/service-context this))]
-     (swap! environment-class-info assoc env-name tag)))
+     (swap! environment-class-info
+            #(let [cache-last-updated
+                   (get-in % [env-name :last-updated])]
+              (if (= cache-last-updated last-update-before-tag-computed)
+                (assoc % env-name
+                         (core/environment-class-info-entry tag))
+                %)))))
```

Comment:
This code is getting big enough to where I'd probably prefer that we try to push it into the core namespace, but, not a blocker for the PR since we are under a time crunch.